README.md
# Modernistik Hyperdrive
Modern application server for building mobile and cloud applications quickly at scale in a service oriented architecture (SOA/Microservices). Hyperdrive is composed of the [Parse Platform server](http://parseplatform.org/) with additional features that allow for best practice deployments. The default configuration allows for easy deployment using [Heroku](https://www.heroku.com/) or [Amazon Web Services (AWS)](https://aws.amazon.com/).

The additional features include:

* Both [MongoDB](https://www.mongodb.com/) and [PostgreSQL](https://www.postgresql.org/) supported as database adapters.
* [AWS S3](https://aws.amazon.com/s3/), Google Cloud Storage or MongoDB GridFS for file storage.
* Routing of incoming webhooks into cloud functions.
* Support for [Mailgun](https://www.mailgun.com/) and [AWS Simple Email Service](https://aws.amazon.com/ses/) (SES) email adapters.
* Push Notifications through Amazon SNS for both iOS and Android.
* Built-in [Parse Dashboard](https://github.com/parse-community/parse-dashboard)
* Simplified microservice configuration with available ENV API.

# Getting Started
To quickly get started with Hyperdrive, you will first need to install it from the [NPM repository](https://nodejs.org/en/).

```bash
$ npm install -g @modernistik/hyperdrive
```

If you need to install a quick MongoDB database, you can do so with `mongodb-runner`.

```bash
# Install mongodb and start the database service
$ npm install -g mongodb-runner

# Start MongoDB localhost:27017
$ mongodb-runner start
```

Once installed and a database is running, you can then start a default Hyperdrive instance:

```bash
# start instance using MongoDB localhost:27017
# see `hyperdrive -h` for usage
$ hyperdrive -s
```

After the server has started successfully, the API server will be available at http://localhost:1337/parse, and you should be able to access the built-in dashboard
at the url http://localhost:1337/dashboard. If you have any connection errors, make sure the database server is available and you can connect to it.

By default, if no AWS S3 configuration is provided, Hyperdrive will store files directly in the database.

Hyperdrive can mostly be configured by setting environment variables which combine specific Hyperdrive variables with the ones used by Parse Server. To see a list of
variables used by Hyperdrive use the `-h` option or the `--info <optName>` to get details.

```bash
# Get help
$ hyperdrive -h

# Get info on a specific configuration
$ hyperdrive --info databaseURI
# databaseURI : DATABASE_URI
#
#   The database url for either MongoDB or PostgreSQL.
```

## Copyright and License
Copyright (C) 2018 Modernistik LLC (https://www.modernistik.com). Requires agreement for commercial use.


CHANGELOG.md
## Modernistik Hyperdrive Changelog

### 0.2.6
* Documenation changes.

### 0.2.3
* Exposes REDIS_URL in config api.
* Fixes cli load path.

### 0.2.2
* Supports generating all keys with the option '--keys'
* Additional Heroku compatibility

### 0.2.0
* Initial release as an NPM module.


lib/index.js
// Copyright (C) 2016 Modernistik LLC. All Rights Reserved.
// Author: Anthony Persaud
const _ = require('lodash');
const bodyParser = require("body-parser");
const url = require('url');
const path = require('path');
const express = require('express');
const Parse = require('parse/node'); // In a node.js environment
const ParseServer = require('parse-server').ParseServer;
const ParseServerVersion = require('parse-server/package.json').version;
const HyperdriveVersion = require('../package.json').version;
const AWS = require('aws-sdk');
const ServerOpts = require('./definitions');
const DefaultRedisURL = "redis://localhost:6379";
const colors = require('colors');

function warn(msg) {
  console.log(`[WARN] ${msg}`.yellow);
}

// returns the express app.
class Hyperdrive {

  constructor(opts) {
    // load items from definitions.
    this._started = false;
    this.parseServerVersion = ParseServerVersion;
    this.config = ServerOpts.defaults();
    this.app = express();
    this.redisURL = null;
    this.awsConfig = null;
    if (_.isObject(opts)) {
      Object.assign(this.config, opts);
    }
    this._postProcessOptions()

  }

  _postProcessOptions() {
    let config = this.config;
    // if server url is not defined, we can create one.
    if( _.isEmpty(config.serverURL) ){
      // if we have a server host, then build the url.
      let baseServerURL = `http://localhost:${config.port}`;

      if(config.serverHost) {
        baseServerURL = `https://${config.serverHost}`;
      } // otherwise, use localhost
      console.log(config.parseMount);
      config.serverURL = `${baseServerURL}${config.parseMount}`;
    }

    this.serverURL = url.parse(config.serverURL);

    if (config.redisURL == "default"){
      config.redisURL = DefaultRedisURL;
    }

    if (_.isEmpty(config.appName)) {
      config.appName = this.serverURL.hostname;
    }

    if (_.isEmpty(config.systemEmailAddress)) {
      config.systemEmailAddress = `no-reply@${this.serverURL.hostname}`
    }

  }

  start() {
    let config = this.config;

    if (this._started) {
      warn("This instance has already been configured!");
      return;
    }

    if (!config.appId || config.appId.length < 5) {
      warn('Application Id should be greater than 5 characters.'.red);
      process.exit();
    }

    if (!config.masterKey || config.masterKey.length < 5) {
      warn('Master key should be greater than 5 characters.'.red);
      process.exit();
    }

    this.configureAWS();
    this.configureRedisCacheAdapter();
    this.configureFileStorage();
    this.configurePushNotifications();
    this.configureEmailAdapter();
    this.configureFacebook();
    const app = this.app;
    // base health-check
    app.get('/', function(req, res) {
      res.status(200).send(ParseServerVersion);
    });

    this.mountMiddlewares(app);
    this.mountStaticFiles(app);
    this.mountAPI(app);
    this.mountDashboard(app);
    this.mountIncoming(app)
    this.mountServerConfig(app)

    if(_.isFunction(config.beforeBoot)){
      config.beforeBoot(app);
    }

    this.displayStartupLog(app)
    const httpServer = require('http').createServer(app);

    if(_.isNil(config.port)){
      console.log("No port defined to start server! Verify PORT environment variable is set.".red);
      process.exit();
    }

    httpServer.listen(config.port, function() {
      console.log("\n");
      this._started = true;
      if (config.startLiveQueryServer) {
        console.log(`[Hyperdrive.${process.pid}] live query server started...`.green);
        ParseServer.createLiveQueryServer(httpServer);
      }
      console.log(`[Hyperdrive.${process.pid}] started ${config.serverURL}...`.green);

      if(_.isFunction(config.afterBoot)){
        config.afterBoot(app);
      }

    });

    return app;
  }

  configureAWS() {
    let config = this.config;
    if (this.accessKeyId && this.secretAccessKey) {

      this.awsConfig = new AWS.Config({
        accessKeyId: config.accessKeyId,
        secretAccessKey: config.secretAccessKey,
        region: _.defaultTo(config.awsRegion, 'us-east-1')
      });

      AWS.config.apiVersions = {
        s3: '2006-03-01'
      };

    }

  }

  isAWSConfigured() {
    return false == _.isEmpty(this.awsConfig);
  }

  isS3Configured() {
    return this.awsConfig && this.fileStorage == "s3";
  }

  configureRedisCacheAdapter() {
    let config = this.config;
    if (_.isEmpty(config.redisURL))
      return;

    const RedisCacheAdapter = require('parse-server').RedisCacheAdapter;
    config.cacheAdapter = new RedisCacheAdapter({url: config.redisURL});
    this.redisURL = config.redisURL;
    Parse.redisURL = this.redisURL;
  }

  configureFileStorage() {
    let config = this.config;
    // if already configured, then return.
    if (config.filesAdapter)
      return;

    if (this.isAWSConfigured() && config.s3Bucket) {
      const S3Adapter = require('parse-server').S3Adapter;
      config.filesAdapter = new S3Adapter(config.s3Bucket, {
        region: config.awsRegion,
        directAccess: config.s3DirectAccess || false,
        bucketPrefix: config.s3BucketPrefix,
        baseUrl: config.s3BaseUrl,
        globalCacheControl: config.s3GlobalCacheControl,
        ServerSideEncryption: config.s3ServerSideEncryption
      });
      this.fileStorage = "s3";
    } else {
      this.fileStorage = "database";
    }

  }

  configurePushNotifications() {
    let config = this.config;
    // if already configured, then return
    if (config.push)
      return;

    // if no AWS config, then return
    if (!this.isAWSConfigured())
      return;

    // perform sns configuration
    const pushConfig = {
      pushTypes: {},
      accessKey: config.accessKeyId,
      secretKey: config.secretAccessKey,
      region: config.awsRegion
    };

    if (config.snsFCM) {
      pushConfig.pushTypes.android = {
        ARN: config.snsFCM
      };
    } else {
      warn("Android notifications not configured.");
    }

    if (config.snsAPNS) {
      pushConfig.pushTypes.ios = {
        ARN: config.snsAPNS,
        production: _.isNil(config.snsAPNS.match('/APNS_SANDBOX')),
        bundleId: config.iosBundleId
      }
    } else {
      warn("APNS not configured.");
    }

    const SNSPushAdapter = require('parse-server-sns-adapter');
    config.push = {
      adapter: new SNSPushAdapter(pushConfig)
    };

  }

  configureEmailAdapter() {
    let config = this.config;
    // if already configured, then return.
    if (config.emailAdapter)
      return;

    if (config.mailgunAPIKey) {
      config.emailAdapter = {
        module: '@parse/simple-mailgun-adapter',
        options: {
          fromAddress: config.systemEmailAddress,
          apiKey: config.mailgunAPIKey,
          domain: config.mailgunDomain || this.serverURL.hostname
        }
      };
    } else if (this.isAWSConfigured()) {
      config.emailAdapter = {
        module: "parse-server-amazon-ses-adapter",
        options: {
          from: config.systemEmailAddress,
          accessKeyId: config.accessKeyId,
          secretAccessKey: config.secretAccessKey,
          region: config.awsRegion
        }
      }

    } else {
      warn("No email adapter was configured.");
    }

  }

  configureFacebook() {
    let config = this.config;
    if( ! _.isArray(config.facebookAppIds) ) return;

    if( _.isEmpty(config.oauth) ) {
      config.oauth = { facebook: { } };
    }

    config.oauth.facebook = {
      appIds: config.facebookAppIds
    }

  }

  /// MOUNTING
  mountMiddlewares(app) {
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({extended: true}));
  }

  mountStaticFiles(app) {

    let staticPath = this.config.staticFilesPath;
    if (false == _.isEmpty(staticPath)) {
      // ex. '/public'
      app.use(staticPath, express.static(path.join(process.cwd(), staticPath)));
    }
  }

  mountAPI(app) {
    let config = this.config;
    if(_.isEmpty(config.parseMount) || config.parseMount == '-') return;
    Parse.initialize(config.appId, config.javascriptKey, config.masterKey);
    Parse.serverURL = config.serverURL;
    const api = new ParseServer(config);
    app.use(config.parseMount, api);
  }

  mountDashboard(app) {
    let config = this.config;
    if (_.isEmpty(config.dashboardMount) || config.dashboardMount == "-" )
      return;

    const dashboardOpts = {
      "apps": [
        {
          "serverURL": config.serverURL,
          "appId": config.appId,
          "masterKey": config.masterKey,
          "javascriptKey": config.javascriptKey,
          "appName": _.defaultTo(config.appName, "Hyperdrive App")
        }
      ]
    };

    if (config.dashboardUser && app.get('env') != 'development') {
      dashboardOpts.users = [
        {
          "user": config.dashboardUser,
          "pass": config.dashboardPassword
        }
      ];
    }

    const ParseDashboard = require('parse-dashboard');
    const dashboard = new ParseDashboard(dashboardOpts, {allowInsecureHTTP: true});
    app.use(config.dashboardMount, dashboard);

  }

  mountIncoming(app) {
    let config = this.config;
    if (_.isEmpty(config.incomingMount) || config.incomingMount == '-' )
      return;

    app.post(config.incomingMount + '/:method', function(req, res) {
      const method = req.params.method;
      console.log(`[Incoming Webhook Received] ${method} : ${JSON.stringify(req.body)}`);
      Parse.Cloud.run(method, {
        method: method,
        parameters: req.params,
        query: req.query,
        body: req.body
      }).then(function(result) {
        res.send(result);
      }, function(error) {
        console.log(`[Incoming Webhook Received] Failed to route notification ${method} : ${JSON.stringify(error)}`);
        res.status(400).send(error);
      });

    }); // app.post

  }

  mountServerConfig(app) {
    let config = this.config;

    if(config.configMount == '-') return;
    let configMount = config.configMount || config.masterKey;
    app.get('/' + configMount, function(req, res) {

      const parseEnv = {
        PARSE_APPLICATION_ID: config.appId,
        PARSE_SERVER_APPLICATION_ID: config.appId,
        PARSE_CLIENT_KEY: config.clientKey,
        PARSE_SERVER_CLIENT_KEY: config.clientKey,
        PARSE_SERVER_JAVASCRIPT_KEY: config.javascriptKey,
        PARSE_SERVER_MASTER_KEY: config.masterKey,
        PARSE_SERVER_REST_API_KEY: config.restAPIKey,
        PARSE_SERVER_URL: config.serverURL,
        PARSE_SERVER_VERSION: ParseServerVersion,
        PARSE_DATABASE_URI: config.databaseURI,
        DATABASE_URI: config.databaseURI,
        PARSE_SERVER_WEBHOOK_KEY: config.webhookKey,
        AWS_ACCESS_KEY_ID: config.accessKeyId,
        AWS_SECRET_ACCESS_KEY: config.secretAccessKey,
        AWS_REGION: config.awsRegion,
        REDIS_URL: config.redisURL
      };

      if (_.isArray(config.publishEnvKeys) && config.publishEnvKeys.length > 0) {
        let e = process.env;
        config.publishEnvKeys.forEach((key) => {
          if (e[key])
            parseEnv[key] = e[key];
          }
        );
      }
      res.status(200).send(parseEnv);

    }); // app.get
  }

  displayStartupLog(app) {
    let config = this.config;
    let mode = app.get('env');

    console.log(`\n------------ v${HyperdriveVersion}[${ParseServerVersion}] ------------`.green);
    console.log(`ENV           : ${mode}`);
    console.log(`SERVER URL    : ${config.serverURL}`);
    console.log(`DATABASE URI  : ${config.databaseURI}`);
    console.log(`APP ID        : ${config.appId}`);

    if(mode == "development"){
      console.log(`MASTER KEY    : ${config.masterKey}`);
    } else {
      console.log(`MASTER KEY    : ${config.masterKey.substr(0, 5)}**REDACTED**`);
    }

    console.log(`REST API KEY  : ${config.restAPIKey}`);
    console.log(`CLIENT KEY    : ${config.clientKey}`);
    console.log(`JAVASCRIPT KEY: ${config.javascriptKey}`);
    console.log(`WEBHOOKS KEY  : ${config.webhookKey}`);

    console.log(`CLOUD CODE    : ${_.defaultTo(config.cloud,'-')}`);
    console.log(`INCOMING PATH : ${_.defaultTo(config.incomingMount, '-')}`);
    console.log(`CONFIG ROUTE  : ${_.defaultTo(config.configMount, '-')}`);
    console.log(`DASHBOARD     : ${_.defaultTo(config.dashboardMount, '-')}`);
    console.log(`REDIS URL     : ${_.defaultTo(config.redisURL, '-')}`);

    if (config.emailAdapter && config.systemEmailAddress) {
      console.log(`SYSTEM EMAIL  : ${config.systemEmailAddress}`)
    }

    if (this.isAWSConfigured() && this.isS3Configured()) {
      let fileStore = `s3:${config.s3Bucket}/${config.s3BucketPrefix}`;
      console.log(`FILES STORE   : ${fileStore}`);
      console.log(`AWS BASE URL  : /${config.s3BaseURL}`);
    } else {
      console.log(`FILES STORE   : Database`);
    }

    if (this.isAWSConfigured()) {
      console.log(`AWS ACCESS KEY: ${config.accessKeyId}:${config.awsRegion}`);

      if (config.snsAPNS) {
        console.log(`APNS BUNDLE ID : ${config.iosBundleId}`);
        console.log(`APNS SNS       : ${config.snsAPNS}`);
      }

      if (config.snsFCM) {
        console.log(`Firebase SNS   : ${config.snsFCM}`);
      }

      if(_.isArray(config.facebookAppIds)) {
        console.log('Facebook Ids   : ' + _.join(config.facebookAppIds,','));
      }

    }

  }

}

module.exports = Hyperdrive;


lib/parsers.js
// Copyright (C) 2016 Modernistik LLC. All Rights Reserved.
// Author: Anthony Persaud
const path = require('path');
const _ = require('lodash');
// utilize maintained parsers by the Parse-Server team.
/// https://github.com/parse-community/parse-server/blob/master/src/Options/parsers.js
let ParseServerParsers = require('parse-server/lib/Options/parsers.js');

function mountPathOptionalParser(str){
  // if it's false, then return null
  if ( _.isEmpty(str) || str == '-' || str == 'false') {
    return undefined;
  }
  return removeTrailingSlash(str);
}

function removeTrailingSlash(str) {
  if (!str) {
    return str;
  }
  if (str.endsWith("/")) {
    str = str.substr(0, str.length - 1);
  }
  return str;
}

function appendPath(str) {
  let endPath = removeTrailingSlash(str);
  return path.join(process.cwd(), endPath + "/");
}

ParseServerParsers.mountPathOptionalParser = mountPathOptionalParser;
ParseServerParsers.removeTrailingSlash = removeTrailingSlash;
ParseServerParsers.appendPath = appendPath;

module.exports = ParseServerParsers;


lib/cli/hyperdrive.js
#!/usr/bin/env node
// Copyright (C) 2016 Modernistik LLC. All Rights Reserved.
// Author: Anthony Persaud
const _ = require('lodash');
const fs = require('fs');
const path = require('path');
const program = require('commander');
const crypto = require('crypto');
const CURR_DIR = process.cwd();
const Hyperdrive = require('../index.js');
const colors = require('colors');
const HyperdriveVersion = require(path.join(__dirname,'..','..','package.json')).version;
const ParseServerVersion = require('parse-server/package.json').version;
const OptionDefinitions = require('../definitions').definitions;
require('dotenv').config({silent: true});

let versionString = `Hyperdrive: ${HyperdriveVersion} [${ParseServerVersion}]`;
program
  .version(versionString, '-v, --version')
  .usage('[options]')
  .option('-i, --init <name>', 'Initialize a new Hyperdrive project.')
  .option('-s, --start', 'Start the server with default options.')
  .option('--info <optName>', 'Get info on a configuration option.')
  .option('--keys', 'Generate all keys necessary for setup.')
  .option('-k, --key', 'Generate a random secure key to use with the server')
  .option('-c, --config <file>', 'Start a Hyperdrive instance with the configuration')
  .option('-d, --database <databaseURI>', 'The database to use (ex. mongodb://localhost:27017/parse).');

program.on('--help', function(){

    console.log("\n  Examples:");
    console.log('');
    console.log('    $ hyperdrive -s');
    console.log('    $ hyperdrive -c config.js');
    console.log('    $ hyperdrive -d postgres://user@localhost:5432');

    console.log("\n  Configuration : Environment Variables:\n".yellow);
    Object.keys(OptionDefinitions).sort().forEach(function(key){
      let v = OptionDefinitions[key];
      console.log(`    ${key.padEnd(22)}: ${v.env}`);
    });
    console.log("\n  Use `hyperdrive --info <configName>` for additional information".yellow);
    console.log("  Also see `parse-server -h` for additional environment variables.");
    console.log("\n");

  });
program.parse(process.argv);

if(program.info){
  const key = program.info;
  const info = OptionDefinitions[key];
  console.log(`\n${key.green} : ${info.env}`);
  console.log(`\n   ${info.help.green || "No description."}\n`);
  process.exit();
}

if(program.init) {
  const templatePath = `${__dirname}/templates`;
  const projectName = program.init;
  const projectDirectory = path.join(CURR_DIR, projectName);
  if (fs.existsSync(projectDirectory)) {
    console.log('Directory already exists!'.red);
    process.exit();
  }
  console.log(`Creating project: ${projectName}`.green);
  console.log(`=> ${projectDirectory.yellow}`);
  fs.mkdirSync(projectDirectory);
  createDirectoryContents(templatePath, projectDirectory);
  console.log("=> Completed".green)
  console.log(`==> Run 'npm install' in the directory ${projectName}`.green)
  process.exit();

} else if(program.keys) {
  let keys = [
    'PARSE_SERVER_APPLICATION_ID',
    'PARSE_SERVER_MASTER_KEY',
    'PARSE_SERVER_REST_API_KEY',
    'PARSE_SERVER_CLIENT_KEY',
    'PARSE_SERVER_JAVASCRIPT_KEY',
    'PARSE_SERVER_WEBHOOK_KEY'
  ];

  _.forEach(keys, function(value){
    let k = crypto.randomBytes(16).toString('hex').green;
    console.log( `${value.padEnd(30)} = ${k}`);
  });
} else if(program.key) {
  console.log('=>  ' + crypto.randomBytes(16).toString('hex').green);
  process.exit();
} else if(program.start || program.database || program.config) {
  let opts = {};

  if(program.config) {
    let configFile = path.join(CURR_DIR, program.config);
    opts = require(configFile);
  }

  if(program.database){
    opts.databaseURI = program.database;
  }

  const server = new Hyperdrive(opts);
  server.start();
} else {
  program.help(function(t){ return colors.yellow(t); })
}



function createDirectoryContents(templatePath, projectDirectory) {
  const filesToCreate = fs.readdirSync(templatePath);

  filesToCreate.forEach(file => {
    const origFilePath = path.join(templatePath,file);
    // get stats about the current file
    const stats = fs.statSync(origFilePath);

    if (stats.isFile()) {
      const contents = fs.readFileSync(origFilePath, 'utf8');
      const writePath = path.join(projectDirectory,file);
      console.log(`[+]  ${file.yellow}`);
      fs.writeFileSync(writePath, contents, 'utf8');
    }
    else if (stats.isDirectory()) {
      const writePath = path.join(projectDirectory,file);
      console.log(`==>  ${file.yellow}/`);
      fs.mkdirSync(writePath);
      // recursive call
      const fromDir = path.join(templatePath, file);
      const toDir = path.join(projectDirectory,file);
      createDirectoryContents(fromDir,toDir);
    }
  });
}


lib/cli/templates/config.js
const opts = module.exports = {};

// Override options here, ex:
// opts.appName = "My App";
opts.cloud = "./cloud/main.js";

// You can export additional environment variables in the config api.
// opts.publishEnvKeys = ['REDIS_URL'];

// Hook before the server boots
opts.beforeBoot = function(app) {
  // `app` is an expressjs instance
};

// Hook after the server boots
opts.afterBoot = function(app) {
  // `app` is an expressjs instance
};

/* During development, it is useful to have a .env file to set the following
environment variables:

DATABASE_URI                 =
PARSE_SERVER_APPLICATION_ID  =
PARSE_SERVER_MASTER_KEY      =
PARSE_SERVER_REST_API_KEY    =
PARSE_SERVER_CLIENT_KEY      =
PARSE_SERVER_JAVASCRIPT_KEY  =
PARSE_SERVER_WEBHOOK_KEY     =
AWS_ACCESS_KEY_ID            =
AWS_SECRET_ACCESS_KEY        =
S3_BUCKET                    =
PARSE_SERVER_URL             =

*/


lib/definitions.js
// Copyright (C) 2016 Modernistik LLC. All Rights Reserved.
// Author: Anthony Persaud
const _ = require('lodash');
const parsers = require('./parsers');
const crypto = require('crypto');
const url = require('url');
const randomGenKey = crypto.randomBytes(10).toString('hex');
const DefaultRedisURL = "redis://localhost:6379";
const DefaultMongoDB = 'mongodb://localhost:27017/parse';

const ServerDefaults = {
  appName: {
    env: "PARSE_SERVER_APP_NAME",
    default: defaultAppName,
    help: "The Application name."
  },
  appId: {
    env: "PARSE_SERVER_APPLICATION_ID",
    default: crypto.randomBytes(10).toString('hex'),
    help: "The Parse Application ID"
  },
  masterKey: {
    env: "PARSE_SERVER_MASTER_KEY",
    default: crypto.randomBytes(12).toString('hex'),
    help: "The Parse Master Key"
  },
  restAPIKey: {
    env: "PARSE_SERVER_REST_API_KEY",
    default: randomGenKey,
    help: "The key for REST clients to use."
  },
  clientKey: {
    env: "PARSE_SERVER_CLIENT_KEY",
    default: randomGenKey,
    help: "The key for mobile clients to use."
  },
  javascriptKey: {
    env: "PARSE_SERVER_JAVASCRIPT_KEY",
    default: randomGenKey,
    help: "The key for web clients to use."
  },
  webhookKey: {
    env: "PARSE_SERVER_WEBHOOK_KEY",
    help: "The key to use when authenticating cloud code webhook requests.",
    default: randomGenKey
  },
  parseMount: {
    env: "PARSE_SERVER_MOUNT_PATH",
    default: "/parse",
    filter: parsers.mountPathOptionalParser,
    help: "The mount path for the Parse Server."
  },
  serverHost: {
    env: "SERVER_HOSTNAME",
    default: defaultServerHost,
    help: "The public server hostname (ex. myapp.server.com)."
  },
  serverURL: {
    env: "PARSE_SERVER_URL",
    help: "The publicly accessible server url for the api mount point (ex. http://localhost:1337/parse)"
  },
  databaseURI: {
    env: "DATABASE_URI",
    default: defaultDatabaseURI,
    help: "The database url for either MongoDB or PostgreSQL."
  },
  cloud: {
    env: "PARSE_SERVER_CLOUD",
    filter: parsers.appendPath,
    help: "The relative path for the cloudcode main.js file. (ex. ./cloud/main.js)"
  },
  publicServerURL: {
    env: "PARSE_PUBLIC_SERVER_URL",
    default: defaultPublicServerURL,
    help: "The publicly accessible URL for this Parse server. Required for passowrd reset."
  },
  startLiveQueryServer: {
    env: "PARSE_SERVER_START_LIVE_QUERY_SERVER",
    default: false,
    help: "Enables the live query server."
  },
  cluster: {
    env: "PARSE_SERVER_CLUSTER",
    filter: parsers.numberOrBoolParser('cluster'),
    default: true,
    help: "Run with cluster, optionally set the number of processes default to os.cpus().length"
  },
  // Dashboard
  dashboardMount: {
    env: "DASHBOARD_MOUNT",
    default: "/dashboard",
    filter: parsers.mountPathOptionalParser,
    help: "The mount path to access the dashboard. Set to '-' to disable."
  },
  dashboardUser: {
    env: "DASHBOARD_USER",
    default: "admin",
    help: "The username to login to the dashboard."
  },
  dashboardPassword: {
    env: "DASHBOARD_PASSWORD",
    default: "admin",
    help: "The password to login to the dashboard."
  },
  incomingMount: {
    env: "INCOMING_MOUNT",
    default: "/incoming",
    filter: parsers.mountPathOptionalParser,
    help: "The mount location for incoming webhooks to be routed as cloud code functions. Set to '-' to disable."
  },
  configKey: {
    env: "CONFIG_KEY",
    help: "A special psuedorandom key to be used to exposing the server configuration through mountPath '/CONFIG_KEY'. Set to '-' to disable."
  },
  // Server Config
  port: {
    env: "PORT",
    filter: parsers.numberParser('port'),
    default: "1337",
    help: "The port the server should listen on."
  },
  redisURL: {
    env: "REDIS_URL",
    help: "The redis server URL to enable caching and possible job queues."
  },
  verbose: {
    env: "VERBOSE",
    filter: parsers.booleanParser,
    default: false,
    help: "Whether to increase verbosity on Parse Server."
  },
  mailgunAPIKey: {
    env: "MAILGUN_API_KEY",
    help: "Your Mailgun API key."
  },
  mailgunDomain: {
    env: "MAILGUN_DOMAIN",
    help: "The domain setup in Mailgun."
  },
  // AWS Config
  accessKeyId: {
    env: "AWS_ACCESS_KEY_ID",
    help: "Set the AWS access key for the IAM Role that supports the services."
  },
  secretAccessKey: {
    env: "AWS_SECRET_ACCESS_KEY",
    help: "Set the AWS secret key for the IAM Role that supports the services."
  },
  awsRegion: {
    env: "AWS_REGION",
    default: "us-east-1",
    help: "The AWS region where the services are hosted."
  },
  s3Bucket: {
    env: "S3_BUCKET",
    help: "The name of the S3 bucket to store Parse files."
  },
  s3BucketPrefix: {
    env: "S3_BUCKET_PREFIX",
    help: "The prefix to where the files will be stored and located. If defined, it should end with a slash."
  },
  s3BaseURL: {
    env: "S3_BASE_URL",
    filter: parsers.removeTrailingSlash,
    help: "Define the S3 base URL for accessing files. While normally this is an S3 URL, it can be a CloudFront defined URL."
  },
  s3DirectAccess: {
    env: "S3_DIRECT_ACCESS",
    filter: parsers.booleanParser,
    help: "Whether file access should bypass the server.",
    default: false
  },
  s3ServerSideEncryption: {
    env: "S3_SERVER_SIDE_ENCRYPTION",
    help: "ex. AES256|aws:kms"
  },
  s3GlobalCacheControl: {
    env: "S3_GLOBAL_CACHE_CONTROL",
    help: "Set the default cache policy for files on S3.",
    default: "public, max-age=31536000"
  },
  snsAPNS: {
    env: "AWS_SNS_APNS",
    help: "The AWS SNS ARN to be used for sending push notifications to Apple devices."
  },
  snsFCM: {
    env: "AWS_SNS_FCM",
    help: "The AWS SNS ARN to be usd for sending push notifications to Android devices."
  },
  iosBundleId: {
    env: "IOS_BUNDLE_ID",
    help: "The iOS bundle identifier for the iOS version of the mobile app."
  },
  facebookAppIds: {
    env: "FACEBOOK_APP_IDS", // array, comman delimited
    filter: parsers.arrayParser,
    help: "To enable Facebook authentication, set a list of comma delimited Facebook App Ids."
  },
  staticFilesPath: {
    env: "STATIC_FILES_PATH",
    help: "Whether Parse Server should serve static files contained in the `/public` directory."
  },
  systemEmailAddress: {
    env: "SYSTEM_EMAIL_ADDRESS",
    help: "The email that will be used when sending email for verifications and password reset (ex. no-reply@example.com)."
  },
  publishEnvKeys: {
    env: "PUBLISH_ENV_KEYS",
    filter: parsers.arrayParser,
    help: "A list of additional ENV variables to publish in the config API (comma separated).",
    default: []
  }
};

module.exports = {
  definitions: ServerDefaults,
  defaults: function() {
    return parseEnvironmentDefaults(ServerDefaults);
  }
};

/// Parses the definition
function parseEnvironmentDefaults(definitions) {
  transformEnvs();

  return Object.keys(definitions).reduce((memo, key) => {

    const def = definitions[key];
    let val = memo[key];
    // if env defined, try getting value from ENV
    if (_.isNil(val) && def.hasOwnProperty('env')) {
      val = process.env[def.env];
    }
    // if it is still nil, then see if we have a default
    if (_.isNil(val) && def.hasOwnProperty('default')) {

      if (_.isFunction(def.default)) {
        val = def.default();
      } else {
        val = def.default;
      }
    }

    if (false == _.isEmpty(val) && def.hasOwnProperty('filter')) {
      val = def.filter(val);
    }
    // If after all that, we have no value, delete the key.
    if (_.isNil(val)) {
      delete memo[key];
    } else {
      memo[key] = val;
    }

    return memo;
  }, {});
}


function transformEnvs() {
  let env = process.env;

  if(env.REDIS_URL == "default"){
    env.REDIS_URL = DefaultRedisURL;
  }

  env.DATABASE_URI = _.defaultTo(env.DATABASE_URI, env.MONGODB_URI || env.DATABASE_URL || env.PARSE_SERVER_DATABASE_URI);

}

function defaultDatabaseURI() {
  return process.env.MONGODB_URI || process.env.DATABASE_URL || process.env.PARSE_SERVER_DATABASE_URI || DefaultMongoDB;
}

function defaultPublicServerURL() {
  return process.env.PARSE_SERVER_URL;
}

function defaultAppName() {
  return process.env.HEROKU_APP_NAME || "Hyperdrive";
}

function defaultServerHost() {
  // easy support for heroku
  const herokuAppName = process.env.HEROKU_APP_NAME;
  if(herokuAppName) {
    return `${herokuAppName}.herokuapp.com`;
  }
  return undefined;
}
